worker/backup_ee.go

```diff
@@ -260,3 +269,429 @@ func ProcessListBackups(ctx context.Context, location string, creds *x.MinioCred
+	if err != nil {
+		return nil, err
+	}
+	glog.V(3).Infof("Backup manifest version: %d", pr.Request.SinceTs)
+
+	eWriter, err := enc.GetWriter(x.WorkerConfig.EncryptionKey, w)
+	if err != nil {
+		return nil, err
+	}
+
+	// Snappy is much faster than gzip compression, even with the BestSpeed
+	// gzip option. In fact, in my experiments, gzip compression caused the
+	// output speed to be ~30 MBps. Snappy can write at ~90 MBps, and overall
+	// the speed is similar to writing uncompressed data on disk.
+	//
+	// These are the times I saw:
+	// Without compression: 7m2s 33GB output.
+	// With snappy: 7m11s 9.5GB output.
+	// With snappy + S3: 7m54s 9.5GB output.
+	cWriter := snappy.NewBufferedWriter(eWriter)
+
+	stream := pr.DB.NewStreamAt(pr.Request.ReadTs)
+	stream.LogPrefix = "Dgraph.Backup"
+	// Ignore versions less than given sinceTs timestamp, or skip older versions of
+	// the given key by returning an empty list.
+	// Do not do this for schema and type keys. Those keys always have a
+	// version of one. They're handled separately.
+	stream.SinceTs = pr.Request.SinceTs
+	stream.Prefix = []byte{x.ByteData}
+
+	var response pb.BackupResponse
+	stream.KeyToList = func(key []byte, itr *badger.Iterator) (*bpb.KVList, error) {
+		tl := pr.threads[itr.ThreadId]
+		tl.alloc = itr.Alloc
+
+		bitr := itr
+		// Use the threadlocal iterator because "itr" has the sinceTs set and
+		// it will not be able to read all the data.
+		if tl.itr != nil {
+			bitr = tl.itr
+			bitr.Seek(key)
+		}
+
+		kvList, dropOp, err := tl.toBackupList(key, bitr)
+		if err != nil {
+			return nil, err
+		}
+		// we don't want to append a nil value to the slice, so need to check.
+		if dropOp != nil {
+			response.DropOperations = append(response.DropOperations, dropOp)
+		}
+		return kvList, nil
+	}
+
+	predMap := make(map[string]struct{})
+	for _, pred := range pr.Request.Predicates {
+		predMap[pred] = struct{}{}
+	}
+	stream.ChooseKey = func(item *badger.Item) bool {
+		parsedKey, err := x.Parse(item.Key())
+		if err != nil {
+			glog.Errorf("error %v while parsing key %v during backup. Skipping...",
+				err, hex.EncodeToString(item.Key()))
+			return false
+		}
+
+		// Do not choose keys that contain parts of a multi-part list. These keys
+		// will be accessed from the main list.
+		if parsedKey.HasStartUid {
+			return false
+		}
+
+		// Skip backing up the schema and type keys. They will be backed up separately.
+		if parsedKey.IsSchema() || parsedKey.IsType() {
+			return false
+		}
+		_, ok := predMap[parsedKey.Attr]
+		return ok
+	}
+
+	var maxVersion uint64
+	stream.Send = func(buf *z.Buffer) error {
+		list, err := badger.BufferToKVList(buf)
+		if err != nil {
+			return err
+		}
+		for _, kv := range list.Kv {
+			if maxVersion < kv.Version {
+				maxVersion = kv.Version
+			}
+		}
+		return writeKVList(list, cWriter)
+	}
+
+	// This is where the execution happens.
+	if err := stream.Orchestrate(ctx); err != nil {
```

Comment:
This function is orchestrating all `backup magic`. I do not see a `map-reduce` paradigm during backup. IMO - we iterate over groups and process each group separately in a different go routine.

Most of the writing happens in badger. Is this correct?

Reply:
correct. No Map reduce during backup